<a href="https://colab.research.google.com/github/jalorenzo/SparkNotebookColab/blob/master/BDF_12_Exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#00 - Configuration of Apache Spark on Collaboratory


###Installing Java, Spark, and Findspark


---


This code installs Apache Spark 3.0.1, Java 8, and [Findspark](https://github.com/minrk/findspark), a library that makes it easy for Python to find Spark.

In [ ]:
!apt-get update
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget  http://apache.osuosl.org/spark/spark-3.3.1/spark-3.3.1-bin-hadoop3.tgz  
!tar xf spark-3.3.1-bin-hadoop3.tgz  
!rm spark-3.3.1-bin-hadoop3.tgz    
!pip install -q findspark

### Set Environment Variables
Set the locations where Spark and Java are installed.

In [1]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark/"
os.environ["DRIVE_DATA"] = "/content/gdrive/My Drive/Enseignement/2022-2023/ING3/HPDA/BigDataFrameworks/data/"

!rm /content/spark
!ln -s /content/spark-3.3.1-bin-hadoop3 /content/spark
!export SPARK_HOME=/content/spark
!export PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
!echo $SPARK_HOME
!env |grep  "DRIVE_DATA"

rm: cannot remove '/content/spark': No such file or directory
ln: failed to create symbolic link '/content/spark': No such file or directory
/content/spark/
DRIVE_DATA=/content/gdrive/My Drive/Enseignement/2022-2023/ING3/HPDA/BigDataFrameworks/data/


### Start a SparkSession
This will start a local Spark session.

In [5]:
!ls

BDF_01_Introduction_to_Apache_Spark.ipynb	  BDF_10_Spark_MLib.ipynb
BDF_02_Basic_operations_on_Spark.ipynb		  BDF_11_Graph_processing.ipynb
BDF_03_Working_with_DataFrames.ipynb		  BDF_12_Exercises.ipynb
BDF_04_Operations_on_DataFrames.ipynb		  LICENSE.md
BDF_05_RDDs_Resilient_Distributed_Datasets.ipynb  README.md
BDF_06_Persistence_and_Partitioning.ipynb	  content
BDF_07_Advanced_concepts.ipynb			  data
BDF_08_Running_Spark_on_a_cluster.ipynb		  hadoop-2.7.3
BDF_09_Spark_Streaming.ipynb			  setup.py


In [1]:
!python -V

#import findspark
#findspark.init()

import os
os.environ["DRIVE_DATA"] = "./data/"

from pyspark import SparkContext
sc = SparkContext.getOrCreate()

# Example: shows the PySpark version
print("PySpark version {0}".format(sc.version))

# Example: parallelise an array and show the 2 first elements
sc.parallelize([2, 3, 4, 5, 6]).cache().take(2)

Python 3.10.9


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/01/26 00:36:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
PySpark version 3.3.0


[2, 3]

In [2]:
from pyspark.sql import SparkSession
# We create a SparkSession object (or we retrieve it if it is already created)
spark = SparkSession \
.builder \
.appName("My application") \
.master("local[4]") \
.getOrCreate()
#.config("spark.driver.memory", "15g") \
# We get the SparkContext
sc = spark.sparkContext

In [ ]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/gdrive')


---


# 12 - Exercises. Final assignment

## Exercise 12.1

Let us extract information from the cite75_99.txt and apat63_99.txt files. Write a script that performs the following operations:

1. From the cite75_99.txt file, obtain the number of citations received by each patent. You must produce a DataFrame with the following format:

| PatentNum | ncitations |
|-----------|------------|
| 3060453   |    3       |
| 3390168   |    6       |
| 3626542   |   18       | 
| 3611507   |    5       |
| 3000113   |    4       |


2. From the apat63_99.txt file, create a DataFrame to show the patent number, its country and the patent year, discarding the rest of fields in the file. The DataFrame produced must have the following format:

|PatentNum |country|Year |
|----------|-------|-----|
| 3070801  | BE    | 1963| 
| 3070802  | US    | 1963| 
| 3070803  | US    | 1963| 
| 3070804  | US    | 1963| 
| 3070805  | US    | 1963|

 
**Requirements**

 - Both DataFrames must be stored in Parquet format with gzip compression. Check the number of partitions of each DataFrame and the number of files gererated.

 - It is **strongly advised** to copy the files from your Drive to a temporal directory in the notebook virtual machine and unzip them there. This will reduce the execution times. See the cell below:
        

In [10]:
!apt-get install -y bzip2

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Suggested packages:
  bzip2-doc
The following NEW packages will be installed:
  bzip2
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 49.3 kB of archives.
After this operation, 125 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian bullseye/main amd64 bzip2 amd64 1.0.8-4 [49.3 kB]
Fetched 49.3 kB in 1s (83.1 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package bzip2.
(Reading database ... 8860 files and directories currently installed.)
Preparing to unpack .../bzip2_1.0.8-4_amd64.deb ...
Unpacking bzip2 (1.0.8-4) ...
Setting up bzip2 (1.0.8-4) ...


In [14]:
!mkdir tmp
!mkdir ./tmp/data/
!cp "$DRIVE_DATA"apat63_99.txt.tar.bz2 "$DRIVE_DATA"cite75_99.txt.tar.bz2 ./tmp/data
#%cd ./tmp/data
!ls ./tmp/data
!tar -jxf ./tmp/data/apat63_99.txt.tar.bz2 -C ./tmp/data/
!tar -jxf ./tmp/data/cite75_99.txt.tar.bz2 -C ./tmp/data/
!rm ./tmp/data/*.tar.bz2
!ls ./tmp/data

mkdir: cannot create directory ‘tmp’: File exists
mkdir: cannot create directory ‘./tmp/data/’: File exists
apat63_99.txt.tar.bz2  cite75_99.txt.tar.bz2
apat63_99.txt  cite75_99.txt


In [45]:
apat63DF = spark.read.option("header", "true").option("delimiter", ",").option("inferSchema", "true").csv("./tmp/data/apat63_99.txt")
#apat63DF = spark.read.text("./tmp/data/apat63_99.txt")
#spark.createDataFrame(apat63DF, schema=)

apat63DF.show(truncate=False)

+-------+-----+-----+-------+-------+-------+--------+-------+------+------+---+------+-----+--------+--------+-------+--------+--------+--------+--------+--------+--------+--------+
|PATENT |GYEAR|GDATE|APPYEAR|COUNTRY|POSTATE|ASSIGNEE|ASSCODE|CLAIMS|NCLASS|CAT|SUBCAT|CMADE|CRECEIVE|RATIOCIT|GENERAL|ORIGINAL|FWDAPLAG|BCKGTLAG|SELFCTUB|SELFCTLB|SECDUPBD|SECDLWBD|
+-------+-----+-----+-------+-------+-------+--------+-------+------+------+---+------+-----+--------+--------+-------+--------+--------+--------+--------+--------+--------+--------+
|3070801|1963 |1096 |null   |BE     |null   |null    |1      |null  |269   |6  |69    |null |1       |null    |0.0    |null    |null    |null    |null    |null    |null    |null    |
|3070802|1963 |1096 |null   |US     |TX     |null    |1      |null  |2     |6  |63    |null |0       |null    |null   |null    |null    |null    |null    |null    |null    |null    |
|3070803|1963 |1096 |null   |US     |IL     |null    |1      |null  |2     |6  |63   

In [46]:
apat63DF.printSchema()

root
 |-- PATENT: integer (nullable = true)
 |-- GYEAR: integer (nullable = true)
 |-- GDATE: integer (nullable = true)
 |-- APPYEAR: integer (nullable = true)
 |-- COUNTRY: string (nullable = true)
 |-- POSTATE: string (nullable = true)
 |-- ASSIGNEE: integer (nullable = true)
 |-- ASSCODE: integer (nullable = true)
 |-- CLAIMS: integer (nullable = true)
 |-- NCLASS: integer (nullable = true)
 |-- CAT: integer (nullable = true)
 |-- SUBCAT: integer (nullable = true)
 |-- CMADE: integer (nullable = true)
 |-- CRECEIVE: integer (nullable = true)
 |-- RATIOCIT: double (nullable = true)
 |-- GENERAL: double (nullable = true)
 |-- ORIGINAL: double (nullable = true)
 |-- FWDAPLAG: double (nullable = true)
 |-- BCKGTLAG: double (nullable = true)
 |-- SELFCTUB: double (nullable = true)
 |-- SELFCTLB: double (nullable = true)
 |-- SECDUPBD: double (nullable = true)
 |-- SECDLWBD: double (nullable = true)



In [47]:
cite75DF = spark.read.option("header", "true").option("delimiter", ",").option("inferSchema", "true").csv("./tmp/data/cite75_99.txt")

cite75DF.show(truncate=False)

+-------+-------+
|CITING |CITED  |
+-------+-------+
|3858241|956203 |
|3858241|1324234|
|3858241|3398406|
|3858241|3557384|
|3858241|3634889|
|3858242|1515701|
|3858242|3319261|
|3858242|3668705|
|3858242|3707004|
|3858243|2949611|
|3858243|3146465|
|3858243|3156927|
|3858243|3221341|
|3858243|3574238|
|3858243|3681785|
|3858243|3684611|
|3858244|14040  |
|3858244|17445  |
|3858244|2211676|
|3858244|2635670|
+-------+-------+
only showing top 20 rows



In [7]:
cite75DF.printSchema()

root
 |-- CITING: integer (nullable = true)
 |-- CITED: integer (nullable = true)



In [84]:
from pyspark.sql import functions as F
newcite75DF = cite75DF.groupby('CITED').agg((cite75DF.CITED).alias('PatentNum'), F.count(cite75DF.CITED).alias('ncitations')).drop('CITED') #select((cite75DF.CITING).alias('PatentNum'), (cite75DF.CITED).alias('ncitations'))
newapat63DF = apat63DF.select((apat63DF.PATENT).alias('PatentNum'), (apat63DF.COUNTRY).alias('country'), (apat63DF.GYEAR).alias('Year'))

In [85]:
newcite75DF.show()

23/01/20 11:13:33 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.


+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|  3060453|         3|
|  3390168|         6|
|  3626542|        18|
|  3611507|         5|
|  3000113|         4|
|  3273281|         1|
|  3550403|         2|
|   228233|         1|
|  3827399|         1|
|  2622583|         1|
|  3381752|        10|
|  1035529|         1|
|  2664273|         6|
|  3289868|        16|
|  3400790|         7|
|  3618714|         8|
|  3297130|         7|
|  1715276|         4|
|    50223|         1|
|   685851|         1|
+---------+----------+
only showing top 20 rows



In [86]:
newcite75DF.where(F.col("PatentNum")=='3060453').show()

+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|  3060453|         3|
+---------+----------+



In [87]:
newapat63DF.show()

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3070801|     BE|1963|
|  3070802|     US|1963|
|  3070803|     US|1963|
|  3070804|     US|1963|
|  3070805|     US|1963|
|  3070806|     US|1963|
|  3070807|     US|1963|
|  3070808|     US|1963|
|  3070809|     US|1963|
|  3070810|     US|1963|
|  3070811|     US|1963|
|  3070812|     US|1963|
|  3070813|     US|1963|
|  3070814|     US|1963|
|  3070815|     US|1963|
|  3070816|     US|1963|
|  3070817|     US|1963|
|  3070818|     US|1963|
|  3070819|     US|1963|
|  3070820|     GB|1963|
+---------+-------+----+
only showing top 20 rows



In [88]:
!rm -r ./tmp/data/newapat63DF.parquet

In [89]:
newapat63DF.write.format("parquet")\
    .mode("overwrite")\
    .save("./tmp/data/newapat63DF.parquet")

In [90]:
!rm -r ./tmp/data/newcite75DF.parquet

In [91]:
newcite75DF.write.format("parquet")\
    .mode("overwrite")\
    .save("./tmp/data/newcite75DF.parquet")

23/01/20 11:15:36 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.


In [92]:
newapat63DF.rdd.getNumPartitions()

2

In [93]:
newcite75DF.rdd.getNumPartitions()

23/01/20 11:16:25 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.


2

In [94]:
!ls ./tmp/data/

apat63_99.txt  cite75_99.txt  newapat63DF.parquet  newcite75DF.parquet


In [95]:
!ls -alh ./tmp/data/newcite75DF.parquet # number of partitions : 2, number of files created : 4

total 16M
drwxrwxrwx 1 1000 1000 4.0K Jan 20 11:15 .
drwxrwxrwx 1 1000 1000 4.0K Jan 20 11:15 ..
-rwxrwxrwx 1 1000 1000    8 Jan 20 11:15 ._SUCCESS.crc
-rwxrwxrwx 1 1000 1000  60K Jan 20 11:15 .part-00000-4e3077c4-1fb5-4ecc-b8ab-5de2053d5e76-c000.snappy.parquet.crc
-rwxrwxrwx 1 1000 1000  60K Jan 20 11:15 .part-00001-4e3077c4-1fb5-4ecc-b8ab-5de2053d5e76-c000.snappy.parquet.crc
-rwxrwxrwx 1 1000 1000    0 Jan 20 11:15 _SUCCESS
-rwxrwxrwx 1 1000 1000 7.5M Jan 20 11:15 part-00000-4e3077c4-1fb5-4ecc-b8ab-5de2053d5e76-c000.snappy.parquet
-rwxrwxrwx 1 1000 1000 7.5M Jan 20 11:15 part-00001-4e3077c4-1fb5-4ecc-b8ab-5de2053d5e76-c000.snappy.parquet


In [96]:
!ls -alh ./tmp/data/newapat63DF.parquet # number of partitions : 2, number of files created : 8

total 14M
drwxrwxrwx 1 1000 1000 4.0K Jan 20 11:15 .
drwxrwxrwx 1 1000 1000 4.0K Jan 20 11:15 ..
-rwxrwxrwx 1 1000 1000    8 Jan 20 11:15 ._SUCCESS.crc
-rwxrwxrwx 1 1000 1000  56K Jan 20 11:14 .part-00000-ae88285f-4c08-418e-9c8b-e1e20b08c801-c000.snappy.parquet.crc
-rwxrwxrwx 1 1000 1000  48K Jan 20 11:14 .part-00001-ae88285f-4c08-418e-9c8b-e1e20b08c801-c000.snappy.parquet.crc
-rwxrwxrwx 1 1000 1000    0 Jan 20 11:15 _SUCCESS
-rwxrwxrwx 1 1000 1000 7.0M Jan 20 11:14 part-00000-ae88285f-4c08-418e-9c8b-e1e20b08c801-c000.snappy.parquet
-rwxrwxrwx 1 1000 1000 6.0M Jan 20 11:14 part-00001-ae88285f-4c08-418e-9c8b-e1e20b08c801-c000.snappy.parquet


## Exercise 12.2

Write a code that, from the Parquet files created in the previous exercise, obtains for each country and for each year: the total number of patents, the total number of citations from those patents, the average number of citations and the maximum number of citations. Compute only those values in which there are any values in both files (*inner join*). In addition, each country must show its whole name, obtained from the *country_codes.txt* file. The final DataFrame must look like this one:


|Country            |Year|PatentsNum |TotalCitations|AvgCitations      |MaxCitations|
|-------------------|----|-----------|--------------|------------------|------------|
|Algeria            |1963|2          |7             |3.5               |4           |
|Algeria            |1968|1          |2             |2.0               |2           |
|Algeria            |1970|1          |2             |2.0               |2           |
|Algeria            |1972|1          |1             |1.0               |1           |
|Algeria            |1977|1          |2             |2.0               |2           |
|Andorra            |1987|1          |3             |3.0               |3           |
|Andorra            |1993|1          |1             |1.0               |1           |
|Andorra            |1998|1          |1             |1.0               |1           |
|Antigua and Barbuda|1978|1          |6             |6.0               |6           |
|Antigua and Barbuda|1979|1          |14            |14.0              |14          |
|Antigua and Barbuda|1991|1          |8             |8.0               |8           |
|Antigua and Barbuda|1994|1          |19            |19.0              |19          |
|Antigua and Barbuda|1995|2          |12            |6.0               |11          |
|Antigua and Barbuda|1996|2          |3             |1.5               |2           |
|Argentina          |1963|14         |35            |2.5               |7           |
|Argentina          |1964|20         |60            |3.0               |8           |
|Argentina          |1965|10         |35            |3.5               |10          |
|Argentina          |1966|16         |44            |2.75              |9           |
|Argentina          |1967|13         |60            |4.615384615384615 |14          |

**Requirements**

- The output DataFrame must be saved in a single CSV file, with a header and without any compression.


In [3]:
ncDF = spark.read\
            .format("parquet")\
            .option("mode", "FAILFAST")\
            .load("./tmp/data/newcite75DF.parquet")
ncDF.cache()

DataFrame[PatentNum: int, ncitations: bigint]

In [29]:
ncDF.orderBy('PatentNum').show()
ncDF.printSchema()

+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|        1|         2|
|       13|         2|
|       24|         1|
|       29|         1|
|       31|         2|
|       42|         3|
|       51|         1|
|       52|         1|
|       53|         1|
|       84|         1|
|       87|         1|
|       92|         1|
|      105|         1|
|      107|         2|
|      115|         1|
|      168|         1|
|      175|         1|
|      178|         1|
|      179|         1|
|      186|         1|
+---------+----------+
only showing top 20 rows

root
 |-- PatentNum: integer (nullable = true)
 |-- ncitations: long (nullable = true)



In [5]:
naDF = spark.read\
            .format("parquet")\
            .option("mode", "FAILFAST")\
            .load("./tmp/data/newapat63DF.parquet")
naDF.cache()

DataFrame[PatentNum: int, country: string, Year: int]

In [13]:
from pyspark.sql import functions as F

naDF.orderBy('country', 'Year').where(F.col("country")=='MA').show()

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3104211|     MA|1963|
|  3087224|     MA|1963|
|  3072682|     MA|1963|
|  3120389|     MA|1964|
|  3153481|     MA|1964|
|  3121441|     MA|1964|
|  3195481|     MA|1965|
|  3172809|     MA|1965|
|  3192085|     MA|1965|
|  3171602|     MA|1965|
|  3180877|     MA|1965|
|  3191938|     MA|1965|
|  3231275|     MA|1966|
|  3254745|     MA|1966|
|  3233363|     MA|1966|
|  3281102|     MA|1966|
|  3229648|     MA|1966|
|  3284304|     MA|1966|
|  3337956|     MA|1967|
|  3342406|     MA|1967|
+---------+-------+----+
only showing top 20 rows



In [14]:
naDF.show()
naDF.printSchema()

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3070801|     BE|1963|
|  3070802|     US|1963|
|  3070803|     US|1963|
|  3070804|     US|1963|
|  3070805|     US|1963|
|  3070806|     US|1963|
|  3070807|     US|1963|
|  3070808|     US|1963|
|  3070809|     US|1963|
|  3070810|     US|1963|
|  3070811|     US|1963|
|  3070812|     US|1963|
|  3070813|     US|1963|
|  3070814|     US|1963|
|  3070815|     US|1963|
|  3070816|     US|1963|
|  3070817|     US|1963|
|  3070818|     US|1963|
|  3070819|     US|1963|
|  3070820|     GB|1963|
+---------+-------+----+
only showing top 20 rows

root
 |-- PatentNum: integer (nullable = true)
 |-- country: string (nullable = true)
 |-- Year: integer (nullable = true)



In [15]:
from pyspark.sql.types import StructType, StructField, StringType
import os
os.environ["DRIVE_DATA"] = "./data/"

schema = StructType([
    StructField("short",StringType(),True),
    StructField("full",StringType(),True)
])

country_codes = spark.read.option("delimiter", "\t").csv(os.environ['DRIVE_DATA'] + 'country_codes.txt', schema)

In [16]:
country_codes.show()

+-----+-------------------+
|short|               full|
+-----+-------------------+
|   AF|        Afghanistan|
|   AX|      Aland Islands|
|   AL|            Albania|
|   DZ|            Algeria|
|   AS|     American Samoa|
|   AD|            Andorra|
|   AO|             Angola|
|   AI|           Anguilla|
|   AQ|         Antarctica|
|   AG|Antigua and Barbuda|
|   AR|          Argentina|
|   AM|            Armenia|
|   AW|              Aruba|
|   AC|   Ascension Island|
|   AU|          Australia|
|   AT|            Austria|
|   AZ|         Azerbaijan|
|   BS|            Bahamas|
|   BH|            Bahrain|
|   BD|         Bangladesh|
+-----+-------------------+
only showing top 20 rows



In [17]:
from pyspark.sql import functions as F
#, F.count(F.col("PatentNum"))
#naDF.join(country_codes, naDF.country == country_codes.short).select(F.col("full").alias("country"), F.col("Year"), F.col("PatentNum")).groupby('Year', 'country').agg({'PatentNum':'count'}).show()#.groupby('Year', 'country').count().where(F.col("country")=='Argentina').show()
newnaDF = naDF.join(country_codes, naDF.country == country_codes.short, "inner").select(F.col("full").alias("country"), F.col("Year"), F.col("PatentNum").alias('PatentNum'))
newnaDF.orderBy('country', 'Year').show()

+-------------------+----+---------+
|            country|Year|PatentNum|
+-------------------+----+---------+
|            Albania|1999|  5992357|
|            Algeria|1963|  3091754|
|            Algeria|1963|  3113948|
|            Algeria|1963|  3070963|
|            Algeria|1965|  3208073|
|            Algeria|1965|  3186382|
|            Algeria|1968|  3393271|
|            Algeria|1970|  3550376|
|            Algeria|1972|  3649173|
|            Algeria|1977|  4057649|
|            Algeria|1998|  5849771|
|            Andorra|1987|  4688621|
|            Andorra|1993|  5193231|
|            Andorra|1995|  5478082|
|            Andorra|1998|  5785566|
|            Andorra|1998|  5765303|
|            Andorra|1999|  5894770|
|           Anguilla|1998|  5765334|
|Antigua and Barbuda|1978|  4126850|
|Antigua and Barbuda|1979|  4172981|
+-------------------+----+---------+
only showing top 20 rows



In [18]:
ncDF.where(F.col("PatentNum")=='3060453').show()

+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|  3060453|         3|
+---------+----------+



In [19]:
ncDF.show()

+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|  3060453|         3|
|  3390168|         6|
|  3626542|        18|
|  3611507|         5|
|  3000113|         4|
|  3273281|         1|
|  3550403|         2|
|   228233|         1|
|  3827399|         1|
|  2622583|         1|
|  3381752|        10|
|  1035529|         1|
|  2664273|         6|
|  3289868|        16|
|  3400790|         7|
|  3618714|         8|
|  3297130|         7|
|  1715276|         4|
|    50223|         1|
|   685851|         1|
+---------+----------+
only showing top 20 rows



In [21]:
gf = newnaDF.join(ncDF, 'PatentNum', 'inner').orderBy('country', 'Year')
gf.show()

+---------+-------------------+----+----------+
|PatentNum|            country|Year|ncitations|
+---------+-------------------+----+----------+
|  3091754|            Algeria|1963|         4|
|  3070963|            Algeria|1963|         3|
|  3393271|            Algeria|1968|         2|
|  3550376|            Algeria|1970|         2|
|  3649173|            Algeria|1972|         1|
|  4057649|            Algeria|1977|         2|
|  4688621|            Andorra|1987|         3|
|  5193231|            Andorra|1993|         1|
|  5765303|            Andorra|1998|         1|
|  4126850|Antigua and Barbuda|1978|         6|
|  4172981|Antigua and Barbuda|1979|        14|
|  5013035|Antigua and Barbuda|1991|         8|
|  5345071|Antigua and Barbuda|1994|        19|
|  5472566|Antigua and Barbuda|1995|         1|
|  5457307|Antigua and Barbuda|1995|        11|
|  5525786|Antigua and Barbuda|1996|         2|
|  5536941|Antigua and Barbuda|1996|         1|
|  3085146|          Argentina|1963|    

In [24]:
#newgf = gf.groupby('country', 'Year').agg({'PatentNum':'count', 'ncitations': 'count', 'ncitations': 'avg'}).select(F.col('country'), F.col('Year'), F.col('count(ncitations)').alias('totalCitations')) #.agg({'': ''}).show() #orderBy('country','Year').show()
#newgf.show()
newgf = gf.groupby('country', 'Year').agg(F.count('PatentNum').alias('PatentsNum'), F.sum('ncitations').alias('totalCitations'), F.avg('ncitations').alias('avgCitations'), F.max('ncitations').alias('maxCitations')) #.select(F.col('country'), F.col('Year'), F.col('count(ncitations)').alias('totalCitations')) #.agg({'': ''}).show() #By('country','Year').show()
newgf.orderBy('country','Year').show()

23/01/20 15:30:08 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.
23/01/20 15:30:08 WARN RowBasedKeyValueBatch: Calling spill() on RowBasedKeyValueBatch. Will not spill but return 0.


+-------------------+----+----------+--------------+-----------------+------------+
|            country|Year|PatentsNum|totalCitations|     avgCitations|maxCitations|
+-------------------+----+----------+--------------+-----------------+------------+
|            Algeria|1963|         2|             7|              3.5|           4|
|            Algeria|1968|         1|             2|              2.0|           2|
|            Algeria|1970|         1|             2|              2.0|           2|
|            Algeria|1972|         1|             1|              1.0|           1|
|            Algeria|1977|         1|             2|              2.0|           2|
|            Andorra|1987|         1|             3|              3.0|           3|
|            Andorra|1993|         1|             1|              1.0|           1|
|            Andorra|1998|         1|             1|              1.0|           1|
|Antigua and Barbuda|1978|         1|             6|              6.0|      

In [26]:
newgf.orderBy('country','Year').where(F.col("country")=='Morocco').show()

+-------+----+----------+--------------+------------------+------------+
|country|Year|PatentsNum|totalCitations|      avgCitations|maxCitations|
+-------+----+----------+--------------+------------------+------------+
|Morocco|1963|         2|             2|               1.0|           1|
|Morocco|1965|         4|            10|               2.5|           4|
|Morocco|1966|         3|            10|3.3333333333333335|           5|
|Morocco|1967|         2|            11|               5.5|          10|
|Morocco|1968|         3|            22| 7.333333333333333|           9|
|Morocco|1970|         2|             9|               4.5|           8|
|Morocco|1972|         2|            22|              11.0|          14|
|Morocco|1973|         2|             7|               3.5|           4|
|Morocco|1977|         5|            36|               7.2|          10|
|Morocco|1983|         1|             3|               3.0|           3|
|Morocco|1984|         2|            13|           

In [30]:
newgf.count()

2530

## Exercise 12.3

From the apat63_99.txt file, obtain the number of patents per country and year **using RDDs** (do not use DataFrames). The resulting RDD must be a key/value RDD in which the key is a country and the value a list of tuples. Each tuple will be composed of a year and the number of patents of the country during that year. In addition, the resulting RDD must be sorted by  the country code and, for each country, values must be sorted by year.

Example of output key/value entry:

    (u'PA', [(u'1963', 2), (u'1964', 2), (u'1965', 1), (u'1966', 1), (u'1970', 1), (u'1971', 1), (u'1972', 6), (u'1974', 3), (u'1975', 5), (u'1976', 3), (u'1977', 2), (u'1978', 2), (u'1980', 2), (u'1982', 1), (u'1983', 1), (u'1985', 2), (u'1986', 1), (u'1987', 2), (u'1988', 1), (u'1990', 1), (u'1991', 2), (u'1993', 1), (u'1995', 1), (u'1996', 1), (u'1999', 1)])

**Requirements:**

- You must remove the double quotation marks from the country code.
- Use 8 partitions to read the apat63_99.txt.bz2 file.


In [32]:
naDF.show()

NameError: name 'naDF' is not defined

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3070801|     BE|1963|
|  3070802|     US|1963|
|  3070803|     US|1963|
|  3070804|     US|1963|
|  3070805|     US|1963|
|  3070806|     US|1963|
|  3070807|     US|1963|
|  3070808|     US|1963|
|  3070809|     US|1963|
|  3070810|     US|1963|
|  3070811|     US|1963|
|  3070812|     US|1963|
|  3070813|     US|1963|
|  3070814|     US|1963|
|  3070815|     US|1963|
|  3070816|     US|1963|
|  3070817|     US|1963|
|  3070818|     US|1963|
|  3070819|     US|1963|
|  3070820|     GB|1963|
+---------+-------+----+
only showing top 20 rows



In [ ]:
naDF = naDF.rdd.repartition(8)

In [41]:
naDF.getNumPartitions()

8

In [89]:
import os
rdd_apat = sc.textFile(os.environ['DRIVE_DATA'] + 'apat63_99.txt')

In [90]:
rdd_apat.getNumPartitions()

7

In [91]:
rdd_apat = rdd_apat.repartition(8)

In [92]:
rdd_apat.getNumPartitions()

8

In [93]:
rdd_apat

MapPartitionsRDD[135] at coalesce at NativeMethodAccessorImpl.java:0

In [94]:
rdd_apat.count()

2923923

In [95]:
rdd_apat.take(3)

['3070820,1963,1096,,"GB","",,2,,12,6,63,,0,,,,,,,,,',
 '3070821,1963,1096,,"US","IL",,2,,15,6,69,,1,,0,,,,,,,',
 '3070822,1963,1096,,"US","NY",,2,,401,1,12,,4,,0.375,,,,,,,']

In [96]:
rdd_apat.first()

'3070820,1963,1096,,"GB","",,2,,12,6,63,,0,,,,,,,,,'

In [97]:
newdata = rdd_apat.map(lambda line: line.split(","))

In [98]:
newdata.take(2)

[['3070820',
  '1963',
  '1096',
  '',
  '"GB"',
  '""',
  '',
  '2',
  '',
  '12',
  '6',
  '63',
  '',
  '0',
  '',
  '',
  '',
  '',
  '',
  '',
  '',
  '',
  ''],
 ['3070821',
  '1963',
  '1096',
  '',
  '"US"',
  '"IL"',
  '',
  '2',
  '',
  '15',
  '6',
  '69',
  '',
  '1',
  '',
  '0',
  '',
  '',
  '',
  '',
  '',
  '',
  '']]

In [99]:
tuples = newdata.map(lambda row: (row[0], row[1], row[4].replace('"', '')))

In [100]:
tuples.take(5)

[('3070820', '1963', 'GB'),
 ('3070821', '1963', 'US'),
 ('3070822', '1963', 'US'),
 ('3070823', '1963', 'US'),
 ('3070824', '1963', 'US')]

In [101]:
grouped_tuples = tuples.groupBy(lambda x: x[2])
grouped_tuples.take(3)

[('CH', <pyspark.resultiterable.ResultIterable at 0x7f33be754a60>),
 ('NO', <pyspark.resultiterable.ResultIterable at 0x7f33cce17970>),
 ('LI', <pyspark.resultiterable.ResultIterable at 0x7f33ccad78b0>)]

In [104]:
import pandas as pd

lst = [('3070820', '1963', 'GB'),
 ('3070821', '1963', 'US'),
 ('3070822', '1963', 'US'),
 ('3070823', '1963', 'US'),
 ('3070824', '1963', 'US')]
groupedbycolumn = 1
list(pd.DataFrame(lst).groupby(groupedbycolumn).agg(lambda x: len(x)).drop(len(lst[0])-1, axis=1).itertuples(name=None))

[('1963', 5)]

In [135]:
def f(x):
    import pandas as pd
    groupedbycolumn = 1
    # groupby date, and drop the additional column:
    out = pd.DataFrame(x).groupby(groupedbycolumn).agg(lambda x: len(x)).drop(2, axis=1)
    return list(out.itertuples(name=None))

In [136]:
final_apat = grouped_tuples.mapValues(f) #.map(lambda x: (x[0], x[1].map(lambda x: x[0]))).take(5)

In [137]:
final_apat.filter(lambda x: x[0] == 'PA').collect()

[('PA',
  [('1963', 2),
   ('1964', 2),
   ('1965', 1),
   ('1966', 1),
   ('1970', 1),
   ('1971', 1),
   ('1972', 6),
   ('1974', 3),
   ('1975', 5),
   ('1976', 3),
   ('1977', 2),
   ('1978', 2),
   ('1980', 2),
   ('1982', 1),
   ('1983', 1),
   ('1985', 2),
   ('1986', 1),
   ('1987', 2),
   ('1988', 1),
   ('1990', 1),
   ('1991', 2),
   ('1993', 1),
   ('1995', 1),
   ('1996', 1),
   ('1999', 1)])]

## Exercise 12.4

From the Parquet files created in Exercise 12.1, create a DataFrame that gives the patent or patents with the higher number of citations per country and year, as well as the average of the number of citations per country and year, and the difference between the maximum and the average values. The resulting DataFrame should look like this:


|Country|Year|PatentNum|max  |average       |diff              |
|-------|----|---------|-----|--------------|------------------|
|AD     |1987|4688621  |3    |3.0           |0.0               |
|AD     |1993|5193231  |1    |1.0           |0.0               |
|AD     |1998|5765303  |1    |1.0           |0.0               |
|AE     |1984|4482959  |5    |5.0           |0.0               |
|AE     |1985|4554981  |14   |14.0          |0.0               |
|AE     |1987|4663181  |3    |3.0           |0.0               |
|AE     |1989|4805221  |7    |5.0           |2.0               |
|AE     |1990|4909321  |2    |2.0           |0.0               |
|AE     |1991|5004552  |3    |2.0           |1.0               |
|AE     |1992|5104556  |4    |4.0           |0.0               |
|AE     |1993|5181569  |8    |8.0           |0.0               |
|AE     |1996|5580125  |1    |1.0           |0.0               |
|AG     |1978|4126850  |6    |6.0           |0.0               |
|AG     |1979|4172981  |14   |14.0          |0.0               |
|AG     |1991|5013035  |8    |8.0           |0.0               |
|AG     |1994|5345071  |19   |19.0          |0.0               |
|AG     |1995|5457307  |11   |6.0           |5.0               |
|AG     |1996|5525786  |2    |1.5           |0.5               |

**Requirements:**

- The DataFrame must be sorted by country code and year.
- Do **NOT** replace the country code by its whole name.
- The output must be saved as a single CSV file, with a header and without any compression.


In [3]:
nciteDF = spark.read\
            .format("parquet")\
            .option("mode", "FAILFAST")\
            .load("./tmp/data/newcite75DF.parquet")
nciteDF.cache()

DataFrame[PatentNum: int, ncitations: bigint]

In [4]:
newapatDF = spark.read\
            .format("parquet")\
            .option("mode", "FAILFAST")\
            .load("./tmp/data/newapat63DF.parquet")
newapatDF.cache()

DataFrame[PatentNum: int, country: string, Year: int]

In [6]:
nciteDF.show(5)

+---------+----------+
|PatentNum|ncitations|
+---------+----------+
|  3060453|         3|
|  3390168|         6|
|  3626542|        18|
|  3611507|         5|
|  3000113|         4|
+---------+----------+
only showing top 5 rows



In [7]:
newapatDF.show(5)

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3070801|     BE|1963|
|  3070802|     US|1963|
|  3070803|     US|1963|
|  3070804|     US|1963|
|  3070805|     US|1963|
+---------+-------+----+
only showing top 5 rows



In [56]:
from pyspark.sql import functions as F

output = newapatDF.join(nciteDF, 'PatentNum', 'inner').groupby('Year').agg(F.first('PatentNum').alias('PatentNum'), F.max('ncitations').alias('maxCitations'), F.avg('ncitations').alias('avgCitations')).orderBy(['country']).withColumn('diffCitations', F.col('maxCitations')-F.col('avgCitations'))
output

DataFrame[country: string, Year: int, PatentNum: int, maxCitations: bigint, avgCitations: double, diffCitations: double]

DataFrame[Year: int, country: string, PatentNum: int, maxCitations: bigint, avgCitations: double, diffCitations: double]

In [62]:
output.show()

+-------+----+---------+------------+------------+-------------+
|country|Year|PatentNum|maxCitations|avgCitations|diffCitations|
+-------+----+---------+------------+------------+-------------+
|     AD|1987|  4688621|           3|         3.0|          0.0|
|     AD|1998|  5765303|           1|         1.0|          0.0|
|     AD|1993|  5193231|           1|         1.0|          0.0|
|     AE|1993|  5181569|           8|         8.0|          0.0|
|     AE|1992|  5104556|           4|         4.0|          0.0|
|     AE|1990|  4909321|           2|         2.0|          0.0|
|     AE|1996|  5580125|           1|         1.0|          0.0|
|     AE|1991|  4997041|           3|         2.0|          1.0|
|     AE|1987|  4663181|           3|         3.0|          0.0|
|     AE|1984|  4482959|           5|         5.0|          0.0|
|     AE|1989|  4805221|           7|         5.0|          2.0|
|     AE|1985|  4554981|          14|        14.0|          0.0|
|     AG|1995|  5457307| 

In [63]:
output.write.option("header",True).csv('./out/output_12-4')

## Exercise 12.5

From the Parquet file with the (PatentNum,Country,Year) information from Exercise 12.1, create a DataFrame that shows the number of patents associated to each country per decade (understanding as a *decade* the years from 0 to 9; e.g. from 1970 to 1979). In addition, the DataFrame must show the increase or decrease of the number of patents per country and decade with respect to the previous decade. The resulting DataFrame must look like this:

|Country|Decade|PatentsNum|Diff|
|-------|------|----------|----|
|AD     |1980  |1         |0   |
|AD     |1990  |5         |4   |
|AE     |1980  |7         |0   |
|AE     |1990  |11        |4   |
|AG     |1970  |2         |0   |
|AG     |1990  |7         |5   |
|AI     |1990  |1         |0   |
|AL     |1990  |1         |0   |
|AM     |1990  |2         |0   |
|AN     |1970  |1         |0   |
|AN     |1980  |2         |1   |
|AN     |1990  |5         |3   |
|AR     |1960  |135       |0   |
|AR     |1970  |239       |104 |
|AR     |1980  |184       |-55 |
|AR     |1990  |292       |108 |

**Requirements**

- The DataFrame must be sorted by country code and year.
- Do **NOT** replace the country code by its whole name.
- The output must be saved as a single CSV file, with a header and without any compression.

In [29]:
newapatDF.show(5)

+---------+-------+----+
|PatentNum|country|Year|
+---------+-------+----+
|  3070801|     BE|1963|
|  3070802|     US|1963|
|  3070803|     US|1963|
|  3070804|     US|1963|
|  3070805|     US|1963|
+---------+-------+----+
only showing top 5 rows



In [64]:
from pyspark.sql import functions as F

grouped_df = newapatDF.groupBy("country", (F.floor((F.col('Year')/10))*10).alias("decade")) # round the year to the nearest decade
patents_per_decade_df = grouped_df.agg(F.count("*").alias("PatentsNum"))
patents_per_decade_df.orderBy(['country', 'decade']).show()

+-------+------+----------+
|country|decade|PatentsNum|
+-------+------+----------+
|     AD|  1980|         1|
|     AD|  1990|         5|
|     AE|  1980|         7|
|     AE|  1990|        11|
|     AG|  1970|         2|
|     AG|  1990|         7|
|     AI|  1990|         1|
|     AL|  1990|         1|
|     AM|  1990|         2|
|     AN|  1970|         1|
|     AN|  1980|         2|
|     AN|  1990|         5|
|     AR|  1960|       135|
|     AR|  1970|       239|
|     AR|  1980|       184|
|     AR|  1990|       292|
|     AT|  1960|       950|
|     AT|  1970|      2588|
|     AT|  1980|      3057|
|     AT|  1990|      3665|
+-------+------+----------+
only showing top 20 rows



In [ ]:
newapatDF.groupby()